modules/ding_place2book/ding_place2book.module

```diff
@@ -442,61 +442,65 @@ function ding_place2book_get_event_state($event_maker_id, $event_id) {
-              }
-              break;
+    $end = strtotime($event->end_at);
+    $available_tickets = $meta_data->available_tickets;
+    $now = time();
+
+    // Try to derive an event status for the event by looking at the p2b event
+    // state, start date, available tickets and ticket types sales periods.
+    // See: http://developer.place2book.com/v1/events/ for a list possible p2b
+    // event states and what they mean.
+    $status = '';
+    // Start out by rulling out some easy conditions looking at the events start
+    // and end date.
+    if ($now >= $start && $now <= $end) {
+      // The event is currently taking place and ticket sales is over.
+      $status = 'closed';
+    }
+    elseif ($now > $end) {
+      // The event has already taking place.
+      $status = 'event-over';
+    }
+    elseif ($event_state === 'passive') {
+      $status = 'passive';
+    }
+    elseif ($event_state === 'published') {
+      // Now we know the event is not passive and is not over/taking place. If
+      // it's published it can now either be open, upcoming, closed or waiting
+      // list. Find out which by looking at the tickets and ticket types.
+      if ($available_tickets > 0 || strval($available_tickets) == '∞') {
+        // We have available tickets, so find out if sale is open or upcoming.
+        $prices = $p2b->getPrices($event_maker_id, $event_id);
+        foreach ($prices as $price) {
+          $price_sale_begin = strtotime($price->sale_begin_at);
+          $price_sale_end = strtotime($price->sale_end_at);
+          if ($now >= $price_sale_begin && $now <= $price_sale_end) {
+            // We have found an open sales period. No need to continue iterating
+            // the ticket types.
+            $status = 'open';
+            break;
           }
-          break;
-
-        case 'pause':
-          // Go through the sales status.
-          switch ($meta_data->sales_status) {
-            case 'upcoming':
-              $status = 'upcoming';
-              break;
-
-            default:
-              $status = 'closed';
-              break;
+          elseif ($now <= $price_sale_begin) {
+            // We haven't found any open sales period yet,q but at least there's
+            // an upcoming, so set that status for now.
+            $status = 'upcoming';
           }
-          break;
-
-        default:
-          $status = 'closed';
-          break;
+        }
+        // If no upcoming or open sales period was found for the published event
+        // we consider the sales status to be closed. In these cases there might
+        // be a waiting list so we look for that before setting status closed.
+        if (empty($status)) {
+          $status = !empty($meta_data->waiting_list) ? 'waiting-list' : 'closed';
+        }
+      }
+      else {
+        $status = 'sold-out';
       }
     }
     else {
-      $status = 'event-over';
-      $class = '';
-
+      $status = 'closed';
```

Comment:
This means that if event state is draft, held, completed it will show “closed”. I don’t like that :-)

Reply:
Nope, because I check the dates in the beginning. So if an event is held/completed for example it should never get here. We can't rely on the held/completed being set apparently, so it's better to just check the dates.

Reply:
Your right about the 'draft' though.

Reply:
But why set anything at all here?

Reply:
If event state is pause, the sale needs to be closed.